File: Back-End/backend/src/models/Bond.ts
import * as admin from 'firebase-admin';
import { db } from '../global/firebase';

// ── 型別定義 ────────────────────────────────────────────────────────────────

export interface BondInput {
  name: string;
  couponRate: number;   // e.g. 0.045 = 4.5%
  maturityDate: string; // YYYY-MM-DD
  currency: string;     // 'USD' | 'TWD' | ...
  faceValue: number;
  note?: string;
}

export interface BondDoc extends Required<BondInput> {
  id: string;
  createdAt: Date;
  updatedAt: Date;
}

const COL = 'bonds';

// ── Model ───────────────────────────────────────────────────────────────────

export class Bond {
  static async findAll(): Promise<BondDoc[]> {
    const snap = await db.collection(COL).orderBy('created_at', 'asc').get();
    return snap.docs.map(d => deserialize(d));
  }

  static async findById(id: string): Promise<BondDoc | null> {
    const doc = await db.collection(COL).doc(id).get();
    return doc.exists ? deserialize(doc) : null;
  }

  static async create(input: BondInput): Promise<BondDoc> {
    const ref = db.collection(COL).doc();
    await ref.set({
      name:          input.name,
      coupon_rate:   input.couponRate,
      maturity_date: input.maturityDate,
      currency:      input.currency.toUpperCase(),
      face_value:    input.faceValue,
      note:          input.note ?? '',
      created_at:    admin.firestore.FieldValue.serverTimestamp(),
      updated_at:    admin.firestore.FieldValue.serverTimestamp(),
    });
    const created = await ref.get();
    return deserialize(created);
  }

  static async update(id: string, input: Partial<BondInput>): Promise<BondDoc | null> {
    const ref = db.collection(COL).doc(id);
    const existing = await ref.get();
    if (!existing.exists) return null;

    const patch: Record<string, unknown> = {
      updated_at: admin.firestore.FieldValue.serverTimestamp(),
    };
    if (input.name          !== undefined) patch.name          = input.name;
    if (input.couponRate    !== undefined) patch.coupon_rate   = input.couponRate;
    if (input.maturityDate  !== undefined) patch.maturity_date = input.maturityDate;
    if (input.currency      !== undefined) patch.currency      = input.currency.toUpperCase();
    if (input.faceValue     !== undefined) patch.face_value    = input.faceValue;
    if (input.note          !== undefined) patch.note          = input.note;

    await ref.update(patch);
    const updated = await ref.get();
    return deserialize(updated);
  }

  static async delete(id: string): Promise<boolean> {
    const ref = db.collection(COL).doc(id);
    const doc = await ref.get();
    if (!doc.exists) return false;
    await ref.delete();
    return true;
  }
}

// ── 反序列化 ─────────────────────────────────────────────────────────────────

function deserialize(doc: admin.firestore.DocumentSnapshot): BondDoc {
  const d = doc.data()!;
  return {
    id:           doc.id,
    name:         d.name,
    couponRate:   d.coupon_rate,
    maturityDate: d.maturity_date,
    currency:     d.currency,
    faceValue:    d.face_value,
    note:         d.note ?? '',
    createdAt:    d.created_at instanceof admin.firestore.Timestamp
      ? d.created_at.toDate()
      : new Date(),
    updatedAt:    d.updated_at instanceof admin.firestore.Timestamp
      ? d.updated_at.toDate()
      : new Date(),
  };
}

File: Back-End/shioaji-service/models/bond.py
"""@deprecated 已由 foreign_asset 取代。保留供 snapshot 計算讀取舊資料。"""
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "bonds"


def _ts_to_iso(ts) -> Optional[str]:
    return ts.isoformat() if hasattr(ts, "isoformat") else None


def _from_doc(doc) -> dict:
    d = doc.to_dict()
    return {
        "id": doc.id,
        "name": d.get("name", ""),
        "couponRate": d.get("coupon_rate", 0),
        "maturityDate": d.get("maturity_date", ""),
        "currency": d.get("currency", ""),
        "faceValue": d.get("face_value", 0),
        "note": d.get("note", ""),
        "createdAt": _ts_to_iso(d.get("created_at")),
        "updatedAt": _ts_to_iso(d.get("updated_at")),
    }


async def find_all() -> list[dict]:
    def _run():
        snap = get_db().collection(COL).order_by("created_at").get()
        return [_from_doc(d) for d in snap]
    return await asyncio.to_thread(_run)


async def find_by_id(doc_id: str) -> Optional[dict]:
    def _run():
        doc = get_db().collection(COL).document(doc_id).get()
        return _from_doc(doc) if doc.exists else None
    return await asyncio.to_thread(_run)


async def create(data: dict) -> dict:
    def _run():
        ref = get_db().collection(COL).document()
        ref.set({
            "name": data["name"],
            "coupon_rate": data["couponRate"],
            "maturity_date": data["maturityDate"],
            "currency": data["currency"].upper(),
            "face_value": data["faceValue"],
            "note": data.get("note", ""),
            "created_at": SERVER_TIMESTAMP,
            "updated_at": SERVER_TIMESTAMP,
        })
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def update(doc_id: str, patch: dict) -> Optional[dict]:
    def _run():
        ref = get_db().collection(COL).document(doc_id)
        if not ref.get().exists:
            return None
        data: dict = {"updated_at": SERVER_TIMESTAMP}
        if "name"         in patch: data["name"]          = patch["name"]
        if "couponRate"   in patch: data["coupon_rate"]   = patch["couponRate"]
        if "maturityDate" in patch: data["maturity_date"] = patch["maturityDate"]
        if "currency"     in patch: data["currency"]      = patch["currency"].upper()
        if "faceValue"    in patch: data["face_value"]    = patch["faceValue"]
        if "note"         in patch: data["note"]          = patch["note"]
        ref.update(data)
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def delete(doc_id: str) -> bool:
    def _run():
        ref = get_db().collection(COL).document(doc_id)
        if not ref.get().exists:
            return False
        ref.delete()
        return True
    return await asyncio.to_thread(_run)
